protocol/client.go

```diff
@@ -0,0 +1,203 @@
+func (cs *ConiksClient) verifyDirectoryProof(df *DirectoryProof, uname string, key []byte) {
+	ap := df.AP
+	str := df.STR
+	tb := df.TB
+
+	if cs.VerificationResult = cs.verifySTR(str); cs.VerificationResult != Passed {
+		return
+	}
+
+	if cs.ProofType, cs.VerificationResult = cs.verifyAuthPath(uname, key,
+		ap, str); cs.VerificationResult != Passed {
+		return
+	}
+
+	if cs.VerificationResult = cs.verifyReturnedPromise(tb, str, ap); cs.VerificationResult != Passed {
+		return
+	}
+
+	if cs.VerificationResult = cs.verifyIssuedPromises(str, ap); cs.VerificationResult != Passed {
+		return
+	}
+
+	cs.VerificationResult = Passed
+
+	// re-assign new state for the client
+	if str.Epoch > cs.CurrentEpoch {
+		cs.CurrentEpoch++
+		cs.SavedSTR = str.Signature
+	}
+}
+
+func (cs *ConiksClient) verifyDirectoryProofs(dfs *DirectoryProofs, uname string, key []byte) {
+	// TODO: fix me
+	cs.VerificationResult = Passed
+}
+
+func (cs *ConiksClient) verifyAuthPath(uname string, key []byte,
+	ap *m.AuthenticationPath,
+	str *m.SignedTreeRoot) (ProofType, ErrorCode) {
+	proofType := ProofOfAbsence
+
+	// verify VRF Index
+	vrfKey := vrf.PublicKey(str.Policies.VrfPublicKey)
+	if !vrfKey.Verify([]byte(uname), ap.LookupIndex, ap.VrfProof) {
+		return InvalidProof, ErrorBadVRFProof
+	}
+
+	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) { /* proof of inclusion */
+		// verify name-to-key binding
+		// key is nil when the user does lookup for the first time
+		if key != nil && !ap.VerifyBinding(key) {
+			return InvalidProof, ErrorBadBinding
+		}
+		// verify commitment
+		if !ap.Leaf.Commitment.Verify([]byte(uname), ap.Leaf.Value) {
+			return InvalidProof, ErrorBadCommitment
+		}
+		proofType = ProofOfInclusion
+	}
+
+	// verify auth path
+	if !ap.Verify(str.TreeHash) {
+		return InvalidProof, ErrorBadAuthPath
+	}
+
+	return proofType, Passed
+}
+
+func (cs *ConiksClient) verifySTR(str *m.SignedTreeRoot) ErrorCode {
+	// verify STR's signature
+	if !cs.signKey.Verify(str.Serialize(), str.Signature) {
+		return ErrorBadSignature
+	}
+
+	// verify hash chain
+	if (str.Epoch == cs.CurrentEpoch && bytes.Equal(cs.SavedSTR, str.Signature)) ||
+		(str.Epoch == cs.CurrentEpoch+1 && str.VerifyHashChain(cs.SavedSTR)) {
+		return Passed
+	}
+	return ErrorBadSTR
+}
+
+// verifyReturnedPromise verifies the returned TB
+// included in the directory's response.
+func (cs *ConiksClient) verifyReturnedPromise(tb *m.TemporaryBinding,
+	str *m.SignedTreeRoot,
+	ap *m.AuthenticationPath) ErrorCode {
+	if cs.isUseTBs && tb != nil {
+		// verify TB's Signature
+		if !cs.signKey.Verify(tb.Serialize(str.Signature), tb.Signature) {
+			return ErrorBadSignature
+		}
+
+		// verify TB's VRF index
+		if !bytes.Equal(tb.Index, ap.LookupIndex) {
+			return ErrorBadIndex
+		}
+
+		cs.TBs = append(cs.TBs, tb)
+	}
+
+	return Passed
+}
+
+// verifyIssuedPromises verifies issued TBs was inserted
+// in the directory as promised.
+func (cs *ConiksClient) verifyIssuedPromises(str *m.SignedTreeRoot,
```

Comment:
This is just a nit: I would rename this method to `verifyFulfilledPromise` or something similar to make it clearer that this verification happens after the registration (as opposed to `verifyReturnedPromise` which is done before)